src/DurableTask.AzureStorage/Partitioning/AppLeaseManager.cs

```diff
@@ -0,0 +1,272 @@
+        private readonly AppLeaseOptions options;
+        private readonly AzureStorageOrchestrationServiceStats stats;
+
+        private CloudBlobContainer taskHubContainer;
+        private string appId;
+
+        private int isStarted;
+        private bool shutdownComplete;
+        private Task renewTask;
+        private CancellationTokenSource leaseRenewerCancellationTokenSource;
+
+        public AppLeaseManager(string accountName, string taskHub, string workerName, CloudBlobClient storageClient, string taskHubContainerName, string appName, AppLeaseOptions options, AzureStorageOrchestrationServiceStats stats)
+        {
+            this.accountName = accountName;
+            this.taskHub = taskHub;
+            this.workerName = workerName;
+            this.storageClient = storageClient;
+            this.taskHubContainerName = taskHubContainerName;
+            this.appName = appName;
+            this.options = options;
+            this.stats = stats ?? new AzureStorageOrchestrationServiceStats();
+
+            Initialize();
+        }
+
+        void Initialize()
+        {
+            this.taskHubContainer = this.storageClient.GetContainerReference(this.taskHubContainerName);
+
+            using (MD5 md5 = MD5.Create())
+            {
+                byte[] hash = md5.ComputeHash(Encoding.Default.GetBytes(appName));
+                this.appId = new Guid(hash).ToString();
+            }
+        }
+
+        public async Task StartAsync()
+        {
+            if (Interlocked.CompareExchange(ref this.isStarted, 1, 0) != 0)
+            {
+                throw new InvalidOperationException("AppLeaseManager has already started");
+            }
+
+            this.shutdownComplete = false;
+            this.leaseRenewerCancellationTokenSource = new CancellationTokenSource();
+
+            this.renewTask = await Task.Factory.StartNew(() => this.LeaseRenewer());
+        }
+
+        public async Task StopAsync()
+        {
+            if (Interlocked.CompareExchange(ref this.isStarted, 0, 1) != 1)
+            {
+                //idempotent
+                return;
+            }
+
+            await this.ReleaseLeaseAsync();
+            this.shutdownComplete = true;
+
+            if (this.renewTask != null)
+            {
+                this.leaseRenewerCancellationTokenSource.Cancel();
+                await this.renewTask;
+            }
+
+            this.leaseRenewerCancellationTokenSource = null;
+        }
+
+        public async Task<bool> TryAquireAppLeaseAsync()
+        {
+            bool leaseAcquired;
+
+            try
+            {
+                AnalyticsEventSource.Log.AppLeaseAcquisitionStarted(
+                    this.accountName,
+                    this.taskHub,
+                    this.workerName,
+                    this.appId,
+                    Utils.ExtensionVersion);
+
+                appId = await taskHubContainer.AcquireLeaseAsync(this.options.LeaseInterval, appId);
```

Comment:
Is it possible for AcquireLeaseAsync(...) to return a new value for `appId`? I was hoping we could make this instance variable `readonly`.

Reply:
Also, let's prefix member access with `this.`
```suggestion
                await taskHubContainer.AcquireLeaseAsync(this.options.LeaseInterval, this.appId);
```

Reply:
I really don't know. I assume it would always be the leaseId I input here, but my analyzer has made me overly cautious, I suppose.